### ISSUE TITLE
wxPropertyGrid doesn't forward 'wxEVT_COMBOBOX' events when running inside a MFC dialog

### ISSUE THREAD

**Issue by wxtrac** [wxWidgets/wxWidgets#18109]
*Issue migrated from trac ticket # 18109*

**component:** wxPropertyGrid | **priority:** normal | **resolution:** outdated | **keywords:** wxMSW, MFC, closed-due-to-info-needed

#### 2018-03-19 12:19:37: StefanBattmer (Stefan Battmer) created the issue

___
Unfortunately this is rather difficult to reproduce as well as to explain.

This problem exists in wxWidgets 3.1.0 - 3.1.1 but NOT in wxWidgets 3.0.4!

I guess it only happens when running the wxWidgets application inside a windows application. It does NOT occur when compiling a standard wxWidgets app!

What I have is a Windows DLL (a DirectShow interface) that uses the wxPropertyGrid class to display properties of a DirectShow device. Roughly this is done like this:

```cpp
//-----------------------------------------------------------------------------
CwxDialogProp::CwxDialogProp( HWND hwndParent, mvIMPACT::acquire::Device* pDev ) : PropGridFrameBase( wxID_ANY, wxT( "" ), wxDefaultPosition, wxDefaultSize )
//-----------------------------------------------------------------------------
{
    AssociateHandle( hwndParent );
    int width, height;
    GetClientSize( &width, &height );
    m_pPropGridManager = new wxPropertyGridManager( this, widPropertyGridManager, wxDefaultPosition, wxSize( width, height ),
            wxPG_BOLD_MODIFIED | wxPG_SPLITTER_AUTO_CENTER | wxPG_DESCRIPTION | wxPG_TOOLTIPS | wxTAB_TRAVERSAL | wxPGMAN_DEFAULT_STYLE );
    m_pPropGridManager->AddPage( "Device Properties" );
}
```

The 'hwndParent' here is a valid window handle provided by the upper layer.

In general everything is working just fine. The dialog pops up, the grid is displayed. Almost everything can be used. The only thing that DOES NOT work is change events generated by combobox editors that are generated by 'wxEnumProperty' instances. Whenever a selection has been made within the combobox the right event is generated and forwarded into the wxWidgets message handler. All other editors work perfectly! The stack trace then looks like this (wxWidgets 3.1.1 source code):

```
>	DirectShow_acquire.ax!wxAppConsoleBase::AppendPendingEventHandler(wxEvtHandler * toAppend) Line 519	C++
 	DirectShow_acquire.ax!wxEvtHandler::QueueEvent(wxEvent * event) Line 1285	C++
 	DirectShow_acquire.ax!wxEvtHandler::AddPendingEvent(const wxEvent & event) Line 3581	C++
 	DirectShow_acquire.ax!wxVListBoxComboPopup::SendComboBoxEvent(int selection) Line 255	C++
 	DirectShow_acquire.ax!wxVListBoxComboPopup::DismissWithEvent() Line 235	C++
 	DirectShow_acquire.ax!wxVListBoxComboPopup::OnLeftClick(wxMouseEvent & __formal) Line 464	C++
 	DirectShow_acquire.ax!wxAppConsoleBase::HandleEvent(wxEvtHandler * handler, void (wxEvent &) * func, wxEvent & event) Line 657	C++
 	DirectShow_acquire.ax!wxAppConsoleBase::CallEventHandler(wxEvtHandler * handler, wxEventFunctor & functor, wxEvent & event) Line 669	C++
 	DirectShow_acquire.ax!wxEvtHandler::ProcessEventIfMatchesId(const wxEventTableEntryBase & entry, wxEvtHandler * handler, wxEvent & event) Line 1396	C++
 	DirectShow_acquire.ax!wxEventHashTable::HandleEvent(wxEvent & event, wxEvtHandler * self) Line 1004	C++
 	DirectShow_acquire.ax!wxEvtHandler::TryHereOnly(wxEvent & event) Line 1593	C++
 	DirectShow_acquire.ax!wxEvtHandler::TryBeforeAndHere(wxEvent & event) Line 3873	C++
 	DirectShow_acquire.ax!wxEvtHandler::ProcessEventLocally(wxEvent & event) Line 1526	C++
 	DirectShow_acquire.ax!wxEvtHandler::ProcessEvent(wxEvent & event) Line 1499	C++
 	DirectShow_acquire.ax!wxVarScrollHelperEvtHandler::ProcessEvent(wxEvent & event) Line 77	C++
 	DirectShow_acquire.ax!wxEvtHandler::DoTryChain(wxEvent & event) Line 1558	C++
 	DirectShow_acquire.ax!wxEvtHandler::ProcessEventLocally(wxEvent & event) Line 1526	C++
 	DirectShow_acquire.ax!wxEvtHandler::ProcessEvent(wxEvent & event) Line 1499	C++
 	DirectShow_acquire.ax!wxEvtHandler::SafelyProcessEvent(wxEvent & event) Line 1617	C++
 	DirectShow_acquire.ax!wxWindowBase::HandleWindowEvent(wxEvent & event) Line 1540	C++
 	DirectShow_acquire.ax!wxWindow::HandleMouseEvent(unsigned int msg, int x, int y, unsigned int flags) Line 5550	C++
 	DirectShow_acquire.ax!wxWindow::MSWHandleMessage(long * result, unsigned int message, unsigned int wParam, long lParam) Line 3060	C++
 	DirectShow_acquire.ax!wxWindow::MSWWindowProc(unsigned int message, unsigned int wParam, long lParam) Line 3758	C++
 	DirectShow_acquire.ax!wxWndProc(HWND__ * hWnd, unsigned int message, unsigned int wParam, long lParam) Line 2846	C++
 	[External Code]	
 	[Frames below may be incorrect and/or missing, no symbols loaded for user32.dll]	
 	DirectShow_acquire.ax!wxIdleWakeUpModule::MsgHookProc(int nCode, unsigned int wParam, long lParam) Line 7556	C++
 	[External Code]	
 	amcap.exe!0092421f()	Unknown
 	amcap.exe!00937379()	Unknown
 	[External Code]	
 	amcap.exe!0091c381()	Unknown
 	amcap.exe!0094d5cb()	Unknown
 	[External Code]	
 	amcap.exe!0093af8e()	Unknown
 	amcap.exe!0093a150()	Unknown
 	[External Code]	
 	amcap.exe!0091c381()	Unknown
 	amcap.exe!0094d5cb()	Unknown
 	[External Code]	
 	DirectShow_acquire.ax!wxIdleWakeUpModule::MsgHookProc(int nCode, unsigned int wParam, long lParam) Line 7556	C++
 	[External Code]	
 	amcap.exe!009172c0()	Unknown
 	amcap.exe!00950464()	Unknown
 	[External Code]	
```

This is the same when using wxWidgets 3.0.4 even though the sources line numbers might differ. However the event somewhat vanishes then when using the wxWidgets 3.1.x branch resulting in NO 'EVT_PG_CHANGED' handler being called thus the change never gets forwarded to the application. What I would expect is this later:

```
>	DirectShow_acquire.ax!wxPGChoiceEditor::OnEvent(wxPropertyGrid * propGrid, wxPGProperty * property, wxWindow * ctrl, wxEvent & event) Line 1135	C++
 	DirectShow_acquire.ax!wxPropertyGrid::HandleCustomEditorEvent(wxEvent & event) Line 3643	C++
 	DirectShow_acquire.ax!wxPropertyGridEditorEventForwarder::ProcessEvent(wxEvent & event) Line 3909	C++
 	DirectShow_acquire.ax!wxEvtHandler::ProcessPendingEvents() Line 1357	C++
 	DirectShow_acquire.ax!wxAppConsoleBase::ProcessPendingEvents() Line 522	C++
 	DirectShow_acquire.ax!wxIdleWakeUpModule::MsgHookProc(int nCode, unsigned int wParam, long lParam) Line 7496	C++
 	[External Code]	
 	[Frames below may be incorrect and/or missing, no symbols loaded for user32.dll]	
 	amcap.exe!0092421f()	Unknown
 	amcap.exe!00937379()	Unknown
 	[External Code]	
 	amcap.exe!0091c381()	Unknown
 	amcap.exe!0094d5cb()	Unknown
 	[External Code]	
 	amcap.exe!0093af8e()	Unknown
 	amcap.exe!0093a150()	Unknown
 	[External Code]	
 	amcap.exe!0091c381()	Unknown
 	amcap.exe!0094d5cb()	Unknown
 	[External Code]	
 	DirectShow_acquire.ax!wxIdleWakeUpModule::MsgHookProc(int nCode, unsigned int wParam, long lParam) Line 7499	C++
 	[External Code]	
 	amcap.exe!009172c0()	Unknown
 	amcap.exe!00950464()	Unknown
 	[External Code]	
```

But since wxWidgets 3.1.0 this does no longer happen. This stack trace has been generated using the wxWidgets 3.0.4 sources.

Another note: I tried using 'wxEditEnumProperty' instead of 'wxEnumProperty' during tests. Here the same problem exists, but pressing ENTER after the selection generates a change event that is actually reaching the consuming application.

I hope this makes sense and hopefully this rings a bell somewhere! I can provide more information if needed and also a description how to reproduce this on a Windows machine (which would basically result in the installation of 2 software packages(one from me and one from a third party))(both not too large).

**Comment by StefanBattmer** [comment]
Dear @vadz,

I was the original author of this ticket and did just look into it once more. From what I understood now I think my original description is not entirely correct. First of all I think that this issue starts to happen in version 3.1.1 and NOT in 3.1.0 as I stated. I think what actually seems to cause this is the following commit: https://github.com/wxWidgets/wxWidgets/commit/9b4759d7b6dfc1ea016db543405c09ae2f0d49db

Once again: I have a variety of executables all running absolutely fine with new and old versions of wxWidgets. What I am working with right now is version 3.2.2.1.

Apart from that I also have a Windows only DirectShow DLL that embeddeds the wxPropertyGrid into an existing MSW (NOT MFC but I think this doesn't matter here) frame. The way this is done is roughly like this:

`
//-------------------------------------------------------------------------------------------------
BOOL APIENTRY DllMain( HANDLE hModule, DWORD dwReason, LPVOID lpReserved )
//-------------------------------------------------------------------------------------------------
{
    switch( dwReason )
    {
    case DLL_PROCESS_ATTACH:
        wxAppOnInit();
        break;
    case DLL_PROCESS_DETACH:
        wxAppOnExit();
        break;
    }
    return DllEntryPoint( ( HINSTANCE )hModule, dwReason, lpReserved );
}

...

//------------------------------------------------------------------------------
void wxAppOnInit()
//------------------------------------------------------------------------------
{
    {
        int argc = 0;
        wxChar** argv = NULL;
        wxLogNull logSuspendScope; // BAT: To get rid of the 'Error: Cannot initialize OLE' message under LabVIEW
        wxEntryStart( argc, argv );
    }

    if( wxTheApp )
    {
        wxTheApp->CallOnInit();
    }
}

//------------------------------------------------------------------------------
void wxAppOnExit()
//------------------------------------------------------------------------------
{
    if( wxTheApp )
    {
        wxTheApp->OnExit();
    }
    wxEntryCleanup();
}

...

//-----------------------------------------------------------------------------
CwxDialogProp::CwxDialogProp( HWND hwndParent ) : PropGridFrameBase( wxID_ANY, wxT( "" ), wxDefaultPosition, wxDefaultSize ) // this essentially derives from wxFrame!!
//-----------------------------------------------------------------------------
{
    AssociateHandle( hwndParent );
    int width, height;
    GetClientSize( &width, &height );
    m_pPropGridManager = new wxPropertyGridManager( this, widPropertyGridManager, wxDefaultPosition, wxSize( width, height ),
            wxPG_BOLD_MODIFIED | wxPG_SPLITTER_AUTO_CENTER | wxPG_DESCRIPTION | wxPG_TOOLTIPS | wxTAB_TRAVERSAL | wxPGMAN_DEFAULT_STYLE );
 // More Code
}
`

Like described in my old comments **wxEnumProperties** seem to no longer propagate **wxEVT_COMBOBOX** events since wxWidgets 3.1.1. On top of that when using the **wxPG_EDITOR( TextCtrlAndButton )** macro pressing the button also doesn't seem to do anything. 

I think I found the reason for that in the changeset from above: In my DLL based wxWidgets integration 'pending events' no longer get processed thus each time an editor calls **AddPendingEvent** this event is simply pushed into a queue and then remains there since no one on the other side empties these queues. The reason for that is, that in my case in the new **wxApp::MSWProcessPendingEventsIfNeeded** never calls **ProcessPendingEvents** because **wxEventLoop::GetActive()** always returns NULL. Reason for that is, that in my scenario **wxEventLoopBase::SetActive(wxEventLoopBase* loop)** never gets called. Is this somewhat expected? Do I need to call an additional function now? If I simply remove everything except **ProcessPendingEvents** from **wxApp::MSWProcessPendingEventsIfNeeded** everything seems to be back to normal but that obviously not an option...

**Comment by vadz** [comment]
Thanks for debugging this!

It looks like the idea to move wake up handling to the event loop was wrong because even if there is usually one running, it's possible not to have one, as your case shows. So we probably need to move `m_heventWake` to `wxApp` itself, and then we could also check if wake up was requested there. OTOH it's not clear if it's a good to use a single such event for all loops, as now waking up one of them would wake them all up. So perhaps we need to duplicate this part of `wxMSWEventLoopBase` in `wxApp` (which would involve factoring it out in a reusable class, and so would require a bit of work).

Or we could just apply your patch above, of course. The only potential problem with it that I see is that it might be less efficient, as it would call `ProcessPendingEvents()` every time, but OTOH it probably doesn't do that much and might not take much longer than calling `WaitForSingleObject()`. It would be nice to benchmark this to be sure, but this risks being tricky to do.

To summarize: if anything would like to implement the first suggestion, it would be nice. I probably won't be able to do it myself, however, and in this case applying the patch above seems better than doing nothing at all, so let's do this for 3.3.0 if nothing else.

Thanks again!

**Comment by StefanBattmer** [comment]
Thanks for the quick feedback! I can live with what I have right now since you obviously don't dislike it too much so I am happy with any solution at any given time! Thanks again!

**Comment by StefanBattmer** [comment]
@vadz: I just had a look at the code again: I actually don't think that my patch is measurable less efficient than the current state of the code. Only in my _special_ scenario it might run with pre 3.1.1 speed since only then **wxEventLoop::GetActive()** returns NULL. A regular application HAS an active message loop as far as I can see and then the **evtLoop->MSWIsWakeUpRequested()** call still decides if **ProcessPendingEvents** needs to be called.

And when looking at **wxAppConsoleBase::ProcessPendingEvents** I think even calling this function every time doesn't do too much damage. In the end a critical section is acquired and released and inside the locked section a queue would be checked for _emptiness_. So I guess if nothing is fundamentally wrong with the way I changed the code we should be fine with the change performance wise.

**Comment by vadz** [comment]
Sorry for the huge delay. I hoped to find time to implement a "better" (I'm not even sure if it is, but I think it might...) solution, but I'm obviously never going to be able to do it, so I'll just apply your patch above soon.

Thanks!

**Comment by StefanBattmer** [comment]
Great! Thanks!
